storage/index.go

```diff
@@ -207,156 +205,144 @@ func newNamespaceIndexWithOptions(
-		// (0) alloc a slice for the errors in queue, and return a wait group
-		// the slice and an index into the slice.
-		// (1) provide an error channel as input to i.insertQueue.Insert, and
-		// guarantee that receives a value on success/failure in the async
-		// insertion code path. We could eliminate the wait group if we did that.
-		// (2) we could provide an OnIndexError callback within OnIndexSeries,
-		// again ensuring it was called upon failure, and cache the error within
-		// the dbShardEntry. Then we're guaranteed that the error would be set
-		// once wg.Wait returns, and the shard insert code-paths would be able
-		// to retrieve the error from the dbShardEntry. Not pretty, but probably
-		// a lot cheaper than (1).
 		wg.Wait()
+
+		// Resort the batch by initial enqueue order
+		if numErrs := batch.NumErrs(); numErrs > 0 {
+			// Restore the sort order from whene enqueued for the caller
+			batch.SortByEnqueued()
+			return fmt.Errorf("check batch: %d insert errors", numErrs)
+		}
 	}
 
 	return nil
 }
 
-// NB: this function is called by the namespaceIndexInsertQueue.
+// WriteBatches is called by the indexInsertQueue.
 // FOLLOWUP(prateek): propagate error back up from here to the indexInsertQueue
 // so that we can notify users of success/failure correctly in the case of
 // sync'd inserts.
-func (i *nsIndex) writeBatch(inserts []index.WriteBatchEntry) {
+func (i *nsIndex) writeBatches(
+	batches []*index.WriteBatch,
+) {
 	// NB(prateek): we use a read lock to guard against mutation of the
 	// indexBlocks, mutations within the underlying blocks are guarded
 	// by primitives internal to it.
 	i.state.RLock()
 	defer i.state.RUnlock()
-
 	if !i.isOpenWithRLock() {
 		// NB(prateek): deliberately skip calling any of the `OnIndexFinalize` methods
 		// on the provided inserts to terminate quicker during shutdown.
-		i.metrics.InsertAfterClose.Inc(int64(len(inserts)))
+		return
 	}
 
-	// we sort the inserts by which block they're applicable for, and do the inserts
-	// for each block.
-	writesByBlockStart := index.WriteBatchEntryByBlockStart(inserts)
-	sort.Sort(writesByBlockStart)
-	writesByBlockStart.ForEachBlockStart(i.writeBatchForBlockStartWithRLock)
+	now := i.nowFn()
+	futureLimit := now.Add(1 * i.bufferFuture)
+	pastLimit := now.Add(-1 * i.bufferPast)
+	writeBatchFn := i.writeBatchForBlockStartWithRLock
+	for _, batch := range batches {
+		// Ensure timestamp is not too old/new based on retention policies and that
+		// doc is valid.
+		batch.ForEach(func(idx int, entry index.WriteBatchEntry,
+			d doc.Document, _ index.WriteBatchEntryResult) {
+			if !futureLimit.After(entry.Timestamp) {
+				batch.MarkUnmarkedEntryError(m3dberrors.ErrTooFuture, idx)
+				return
+			}
+
+			if !entry.Timestamp.After(pastLimit) {
+				batch.MarkUnmarkedEntryError(m3dberrors.ErrTooPast, idx)
+				return
+			}
+		})
+
+		// we sort the inserts by which block they're applicable for, and do the inserts
```

Comment:
Comment would read better if it just started with "Sort the inserts..."

Reply:
Sure thing.